src/Microsoft.ML.Dnn/TensorTypeExtensions.cs

```diff
@@ -3,49 +3,477 @@
-
             unsafe
             {
-                dst = *(T*)tensor.buffer;
+                if (typeof(T).as_dtype() == tensor.dtype && tensor.dtype != TF_DataType.TF_STRING)
+                {
+                    dst = *(T*) tensor.buffer;
+                    return;
+                }
+
+                //TODO When upgrading to the newest version of Tensorflow.NET, NumSharp will consequently upgrade too, just remove the second argument of the ChangeType calls.
+                switch (tensor.dtype)
+                {
+#if _REGEN
+                    %foreach supported_numericals_TF_DataType,supported_numericals,supported_numericals_lowercase%
+                    case TF_DataType.#1:
+                        dst = Converts.ChangeType<T>(*(#3*) tensor.buffer, NPTypeCode.#2);
+                        return;
+                    %
+#else
+
+                    case TF_DataType.TF_BOOL:
+                        dst = Converts.ChangeType<T>(*(bool*) tensor.buffer, NPTypeCode.Boolean);
+                        return;
+                    case TF_DataType.TF_UINT8:
+                        dst = Converts.ChangeType<T>(*(byte*) tensor.buffer, NPTypeCode.Byte);
+                        return;
+                    case TF_DataType.TF_INT16:
+                        dst = Converts.ChangeType<T>(*(short*) tensor.buffer, NPTypeCode.Int16);
+                        return;
+                    case TF_DataType.TF_UINT16:
+                        dst = Converts.ChangeType<T>(*(ushort*) tensor.buffer, NPTypeCode.UInt16);
+                        return;
+                    case TF_DataType.TF_INT32:
+                        dst = Converts.ChangeType<T>(*(int*) tensor.buffer, NPTypeCode.Int32);
+                        return;
+                    case TF_DataType.TF_UINT32:
+                        dst = Converts.ChangeType<T>(*(uint*) tensor.buffer, NPTypeCode.UInt32);
+                        return;
+                    case TF_DataType.TF_INT64:
+                        dst = Converts.ChangeType<T>(*(long*) tensor.buffer, NPTypeCode.Int64);
+                        return;
+                    case TF_DataType.TF_UINT64:
+                        dst = Converts.ChangeType<T>(*(ulong*) tensor.buffer, NPTypeCode.UInt64);
+                        return;
+                    case TF_DataType.TF_DOUBLE:
+                        dst = Converts.ChangeType<T>(*(double*) tensor.buffer, NPTypeCode.Double);
+                        return;
+                    case TF_DataType.TF_FLOAT:
+                        dst = Converts.ChangeType<T>(*(float*) tensor.buffer, NPTypeCode.Single);
+                        return;
+#endif
+
+                    case TF_DataType.TF_COMPLEX64:
+                    case TF_DataType.TF_COMPLEX128:
+                    default:
+                        throw new NotSupportedException();
+                }
             }
+        }
 
+        public static void ToScalar(this Tensor tensor, ref string dst)
```

Comment:
I don't believe any code is calling this method. Can it be removed?

Reply:
Right, `TensorTypeExtensions` is internal. My bad.